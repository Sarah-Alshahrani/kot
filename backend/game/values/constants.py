import os

# directories
CARDS_DIRECTORY = os.path.dirname(os.path.abspath(__file__))

# ints
DEFAULT_DICE_TO_ROLL = 6
DEFAULT_RE_ROLL_COUNT = 2
DEFAULT_HEALTH = 10
DEATH_HIT_POINT = 0
DEFAULT_ENERGY_CUBE = 0
CARD_STORE_SIZE_LIMITER = 3
ENERGY_HOARDER_DIVIDER = 6
SWEEP_CARD_STORE_COST = 2
VICTORY_POINTS_TO_WIN = 20

# strings
OUT_OF_CARDS_MSG = "No more cards available to shuffle"
INSUFFICIENT_FUNDS_MSG = "Insufficient energy to purchase card"
INSUFFICIENT_FUNDS_TO_SWEEP_MSG = "Insufficient energy to sweep the card store"
