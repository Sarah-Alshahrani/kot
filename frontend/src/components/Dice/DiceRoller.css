.dice_container {
  display: flex;

  flex-direction: column;
  font-family: 'Open Sans', sans-serif;

}