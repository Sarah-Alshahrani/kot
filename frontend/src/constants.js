
// REST Endpoints



export const REST_ENDPOINT_VERSION = 'http://localhost:8000/api/v1/'
export const REST_ENDPOINT_DICE = REST_ENDPOINT_VERSION + 'dice/'

// export const REST_ENDPOINT_DICE = 'http://localhost:8000/api/dice/'