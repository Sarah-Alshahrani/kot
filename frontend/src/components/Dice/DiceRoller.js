import React from "react";
import "./DiceRoller.css";
import Button from "react-bootstrap/Button";

import * as Constants from "../../constants";

import DiceBoard from "./DiceBoard";

import GameInstance from "../../services/gameService";

class DiceRoller extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      username: props.currentUser,
      gameRoom: props.currentRoom,
      rolledDice: [],
      selectedDice: [],
      allowReroll: true
    };

    this.AttemptReroll = this.AttemptReroll.bind(this);
    this.AttemptReroll_via_REST = this.AttemptReroll_via_REST.bind(this);
    GameInstance.addDiceCallback(this.diceRollerHandler.bind(this));
  }

  componentDidMount() {
    // ... do something with fetchedData e.g. set the data
  }

  selectedDiceCallback = selectedDice => {
    this.setState({ selectedDice: selectedDice });
    // console.log('selected :', selectedDice);
    // console.log('total :', this.state.rolledDice.length);
  };

  // sends dice back to server formatted as an array of 2d arrays.
  // first index: the roll value
  // second index: selected or not
  // payload format: [['e', True], ['1', False], ['h', True], ['2', False], ['3', True], ['e', False]]
  sendRollRequest(envelope) {
    GameInstance.sendMessage({
      command: "roll_dice_request",
      user: envelope.user,
      room: envelope.room,
      payload: envelope.number_to_reroll
    });
  }

  diceRollerHandler(message) {
    // const room = message.room;
    // const user = message.user;
    const content = message.content;

    if (content === undefined || content === "") return;

    this.setState({ rolledDice: content });

    return this.state.log;
  }

  async AttemptReroll(/*e*/) {
    try {
      this.determineSelectedDice();

      let rerollThisMany = this.CalculateRerollCount();

      const messageObject = {
        user: this.state.username,
        room: this.state.gameRoom,
        number_to_reroll: rerollThisMany
      };

      this.sendRollRequest(messageObject);
    } catch (exception) {
      console.log(exception);
    }
  }

  async AttemptReroll_via_REST(/*e*/) {
    try {
      this.determineSelectedDice();

      let rerollThisMany = this.CalculateRerollCount();

      let result = await this.RequestDiceRoll(rerollThisMany);

      if (result) {
        this.setState({ rolledDice: result });
      }
    } catch (exception) {
      console.log(exception);
    }
  }

  determineSelectedDice() {
    let selected = this.state.rolledDice;

    if (selected && selected.length > 0) {
      const messageObject = {
        user: this.state.username,
        room: this.state.gameRoom,
        data: selected
      };

      this.sendSelectedDice(messageObject);
      this.setState({
        message: ""
      });
    }
  }

  // sends dice back to server formatted as an array of 2d arrays.
  // first index: the roll value
  // second index: selected or not
  // payload format: [['e', True], ['1', False], ['h', True], ['2', False], ['3', True], ['e', False]]
  sendSelectedDice(envelope) {
    GameInstance.sendMessage({
      command: "selected_dice_request",
      user: envelope.user,
      room: envelope.room,
      payload: envelope.data
    });
  }

  CalculateRerollCount() {
    if (this.state && this.state.selectedDice && this.state.rolledDice) {
      let keptDice = this.state.selectedDice.length;
      let totalDice = this.state.rolledDice.length;
      return totalDice - keptDice;
    }

    return 0;
  }

  async RequestDiceRoll(rerollThisMany) {
    try {
      // TO DO: Provide rerollThisMany to server request

      const res = await fetch(Constants.REST_ENDPOINT_DICE);
      const json_data = await res.json();

      var result = [];

      result.push([json_data.dice1, json_data.dice1_selected]);
      result.push([json_data.dice2, json_data.dice2_selected]);
      result.push([json_data.dice3, json_data.dice3_selected]);
      result.push([json_data.dice4, json_data.dice4_selected]);
      result.push([json_data.dice5, json_data.dice5_selected]);
      result.push([json_data.dice6, json_data.dice6_selected]);

      this.state.allowReroll = json_data.allowReroll;

      return result;
    } catch (err) {
      console.log(err);
    }
  }

  render() {
    return (
      <div className="DiceRoller">
        <DiceBoard
          ref={this.textInput}
          callbackSendDiceSelectionOut={this.selectedDiceCallback}
          data={this.state.rolledDice}
          currentUser={this.state.username}
          currentRoom={this.state.gameRoom}
        />
        <Button
          className="btn btn-secondary"
          disabled={!this.state.allowReroll}
          onClick={this.AttemptReroll}
        >
          Roll
        </Button>
      </div>
    );
  }
}

export default DiceRoller;
