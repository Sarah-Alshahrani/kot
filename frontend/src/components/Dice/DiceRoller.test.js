import React from "react";
import { shallow } from "enzyme";

import * as Constants from "../../constants";

import fetch from "isomorphic-fetch";

import DiceRoller from "./DiceRoller";

describe("Verify DiceRoller layout template", () => {
  it("DiceRoller should render correctly", () => {
    const errorPage = shallow(<DiceRoller />);
    expect(errorPage).toMatchSnapshot();
  });
});

describe("Verify DiceRoller", () => {
  const test_username = "test_user";
  const test_roomname = "test_room";

  it("default state expected to be..", done => {
    const wrapper = shallow(
      <DiceRoller currentUser={test_username} currentRoom={test_roomname} />
    );

    expect(wrapper.state()).toEqual({
      username: test_username,
      gameRoom: test_roomname,
      rolledDice: [],
      selectedDice: [],
      allowReroll: true,
      allowEndTurn: true
    });

    done();
  });

  it("set state expected to..", done => {
    const wrapper = shallow(
      <DiceRoller currentUser={test_username} currentRoom={test_roomname} />
    );

    expect(wrapper.state()).toEqual({
      username: test_username,
      gameRoom: test_roomname,
      rolledDice: [],
      selectedDice: [],
      allowReroll: true,
      allowEndTurn: true
    });

    wrapper.setState(
      {
        username: test_username,
        gameRoom: test_roomname,
        rolledDice: ["", ""],
        selectedDice: ["", "", "", ""],
        allowReroll: false,
        allowEndTurn: true
      },
      () => {
        expect(wrapper.state()).toEqual({
          username: test_username,
          gameRoom: test_roomname,
          rolledDice: ["", ""],
          selectedDice: ["", "", "", ""],
          allowReroll: false,
          allowEndTurn: true
        });

        done();
      }
    );
  });

  it("CalculateRerollCount() expect 2", () => {
    const wrapper = shallow(
      <DiceRoller currentUser={test_username} currentRoom={test_roomname} />
    );

    wrapper.setState({
      username: test_username,
      gameRoom: test_roomname,
      rolledDice: ["", "", "", "", ""],
      selectedDice: ["", "", ""],
      allowReroll: false,
      allowEndTurn: true
    });

    const actual = wrapper.instance().CalculateRerollCount();

    expect(actual).toEqual(2);
  });

  it("selectedDiceCallback(selectedDice) expect state.selectedDice to be set", () => {
    const wrapper = shallow(
      <DiceRoller currentUser={test_username} currentRoom={test_roomname} />
    );

    wrapper.setState({
      username: test_username,
      gameRoom: test_roomname,
      rolledDice: [],
      selectedDice: [],
      allowReroll: true,
      allowEndTurn: true
    });

    const actual = ["", "", "", ""];

    wrapper.instance().selectedDiceCallback(actual);

    expect(wrapper.state()).toEqual({
      username: test_username,
      gameRoom: test_roomname,
      rolledDice: [],
      selectedDice: actual,
      allowReroll: true,
      allowEndTurn: true
    });
  });

  test("should update a state on a next line", () => {
    const wrapper = shallow(
      <DiceRoller currentUser={test_username} currentRoom={test_roomname} />
    );
    wrapper.setState({
      username: test_username,
      gameRoom: test_roomname,
      rolledDice: [],
      selectedDice: [],
      allowReroll: false,
      allowEndTurn: true
    });

    expect(wrapper.state()).toEqual({
      username: test_username,
      gameRoom: test_roomname,
      rolledDice: [],
      selectedDice: [],
      allowReroll: false,
      allowEndTurn: true
    });
  });
});
