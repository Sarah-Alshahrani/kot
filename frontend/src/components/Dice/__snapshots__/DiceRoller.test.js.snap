// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verify DiceRoller layout template DiceRoller should render correctly 1`] = `
<div
  className="DiceRoller"
>
  <DiceBoard
    callbackSendDiceSelectionOut={[Function]}
    data={Array []}
  />
  <Button
    active={false}
    className="btn btn-secondary"
    disabled={false}
    onClick={[Function]}
    type="button"
    variant="primary"
  >
    Roll
  </Button>
  <Button
    active={false}
    className="btn btn-secondary"
    disabled={false}
    onClick={[Function]}
    type="button"
    variant="primary"
  >
    Accept Dice
  </Button>
  <Button
    active={false}
    className="btn btn-secondary"
    disabled={false}
    onClick={[Function]}
    type="button"
    variant="primary"
  >
    End Turn
  </Button>
  <Button
    active={false}
    className="btn btn-secondary"
    disabled={false}
    onClick={[Function]}
    type="button"
    variant="primary"
  >
    Yield Tokyo
  </Button>
</div>
`;
